### PR TITLE
#6 Search Module

### PR DESCRIPTION
**Pull request by hkulekci**
#6 Creating search module 

 - Configuration file created
 - AbstractSearchService added to Search module
 - AbstractIndexService and AbstractQueryService added to Search module
 - Example UserIndexService and UserQueryService added
 - ElasticsearchClient Service and Factory added

### CODE REVIEW

**Comment by hkulekci** [comment]
@edigu ping

**Comment by hkulekci** [comment]
@edigu I already create a basic implementation for Search module. I will try to create a user index and user search method. Right now, user search method create a match_all query on elasticsearch and index method, which is in AbstractIndexService class, directly index data to elasticsearch. Maybe we can add a spesific index method and a search method for user index service and search service.

**Comment by edigu** [comment]
Looks good to me. Can you rebase on master to resolve conflicts? I think search layer can accept/return DTO while indexing & searching data in future. It would be nice improvement.

File: module/Search/src/Search/Exception/MissingDataException.php [hunk 1]
```diff
@@ -0,0 +1,12 @@
+<?php
+/**
+ * Abstract Index Service
+ *
+ * @since     Oct 2015
+ * @author    Haydar KULEKCI <haydarkulekci@gmail.com>
+ */
+namespace Search\Exception;
+
+class MissingDataException extends \InvalidArgumentException
```
Comment (edigu) [inline module/Search/src/Search/Exception/MissingDataException.php]: `MissingArgumentException`?

File: module/Search/src/Search/Query/Service/Factory/UserQueryServiceFactory.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+{
+    /**
+     * Create service
+     *
+     * @param ServiceLocatorInterface $serviceLocator
+     * @return UserIndexService
+     */
+    public function createService(ServiceLocatorInterface $sm)
+    {
+        $userQueryService = new UserQueryService($sm->get('search.client.elastic'));
```
Comment (edigu) [inline module/Search/src/Search/Query/Service/Factory/UserQueryServiceFactory.php]: I would directly return the instance here.. No need to assigning a variable.

File: module/Search/src/Search/Query/Service/AbstractQueryService.php [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+     * With Elastica Query Builder
+     *      $qb = new QueryBuilder();
+     *      $query = $qb->query()->match_all();
+     *      $mainQuery = new \Elastica\Query($query);
+     *      $this->doSearch($mainQuery, 1, 10);
+     *
+     * @param  ElasticaQuery    $query
+     * @return Resultset
+     */
+    public function doSearch(ElasticaQuery $query, $page, $itemsPerPage, $type = self::VIEW_LIST, $version = 1)
```
Comment (edigu) [inline module/Search/src/Search/Query/Service/AbstractQueryService.php]: `protected function doSearch()` would more appropriate.

File: module/Search/src/Search/Exception/MissingDataException.php (hunk 1 above)
Comment (hkulekci) [inline module/Search/src/Search/Exception/MissingDataException.php]: I try to create some exceptions for some exceptional situations. For example, I used that exception on [AbstractIndexService#24](https://github.com/biberlabs/zf2-boilerplate/pull/20/files#diff-5b98d1a21ae8a2a6b123f88f2e938002R24). Maybe we can use same exception on [AbstractSearchService#56](https://github.com/biberlabs/zf2-boilerplate/pull/20/files#diff-236660e217689f195fd9af635637f77bR56) or [AbstractSearchService#87](https://github.com/biberlabs/zf2-boilerplate/pull/20/files#diff-236660e217689f195fd9af635637f77bR87).

File: module/Search/src/Search/Query/Service/Factory/UserQueryServiceFactory.php (hunk 2 above)
Comment (hkulekci) [inline module/Search/src/Search/Query/Service/Factory/UserQueryServiceFactory.php]: :+1:

File: module/Search/src/Search/Query/Service/AbstractQueryService.php (hunk 3 above)
Comment (hkulekci) [inline module/Search/src/Search/Query/Service/AbstractQueryService.php]: :+1:

File: module/Search/src/Search/Exception/MissingDataException.php (hunk 1 above)
Comment (edigu) [inline module/Search/src/Search/Exception/MissingDataException.php]: I understand. In this case `id` is a requirement, part of the data for a valid elastica document, validating incoming data before trying to create documents sounds good. However when searching, its a parameter (argument) which may exists or not. I think separating exceptions to three distinct classes would more appropriate :`MissingArgumentException`, `InvalidDocumentException` and `InvalidArgumentException`

**Pull request merged**